File: pushup_analyzer.py
import json
from exercise_analyzer import ExerciseAnalyzer

class PushupAnalyzer(ExerciseAnalyzer):
    """
    Analyzes a stream of pose landmarks to provide feedback on push-up form.
    Inherits from the abstract ExerciseAnalyzer class.
    """
    def __init__(self):
        super().__init__()
        # --- Push-up Specific State Tracking ---
        self.stage = "up" # Start in the 'up' position
        self.last_feedback = ""
        self.min_angle_in_rep = 180

        # --- Frame Buffer for State Change ---
        self.FRAME_BUFFER = 4
        self.up_frames = 0
        self.down_frames = 0

        # --- Timers for Tempo ---
        self.rep_start_time = 0

        # --- Form Thresholds (Tunable) ---
        self.depth_threshold_angle = 100      # Elbow angle must be less than this for a good rep
        self.rep_threshold = 150              # Single threshold to distinguish up/down phases
        self.body_straight_angle_min = 150    # Angle of shoulder-hip-ankle
        self.elbow_flare_angle_max = 80       # Angle of hip-shoulder-elbow
        self.rep_too_fast_seconds = 1.0

    def process_frame(self, frame_data):
        """
        Processes a single frame of landmark data for push-up analysis.
        """
        self.frame_count += 1
        landmarks = frame_data['landmarks']
        current_time = frame_data['relative_time']
        feedback_this_frame = []

        try:
            # --- Get key landmarks using the helper from the base class ---
            shoulder = self.get_landmark(landmarks, 'right_shoulder')
            elbow = self.get_landmark(landmarks, 'right_elbow')
            wrist = self.get_landmark(landmarks, 'right_wrist')
            hip = self.get_landmark(landmarks, 'right_hip')
            ankle = self.get_landmark(landmarks, 'right_ankle')

            # --- Calculate all relevant angles for this frame ---
            elbow_angle = self.calculate_angle(shoulder, elbow, wrist)
            body_angle = self.calculate_angle(shoulder, hip, ankle)
            elbow_flare_angle = self.calculate_angle(hip, shoulder, elbow)

            # --- 1. Body Straightness Check ---
            if body_angle < self.body_straight_angle_min:
                feedback = "Keep your back straight!"
                if self.last_feedback != feedback:
                    feedback_this_frame.append(feedback)
                    self.last_feedback = feedback
            else:
                self.last_feedback = ""

            # --- 2. Elbow Flare Check ---
            if elbow_flare_angle > self.elbow_flare_angle_max:
                feedback_this_frame.append("Tuck your elbows in a bit!")
            
            # --- 3. Rep Counting, Depth, and State Logic with Frame Buffer ---
            if elbow_angle < self.rep_threshold:
                self.down_frames += 1
                self.up_frames = 0 # Reset the other counter
            else:
                self.up_frames += 1
                self.down_frames = 0 # Reset the other counter
            
            # A rep begins when the user has been 'down' for enough frames.
            if self.stage == 'up' and self.down_frames >= self.FRAME_BUFFER:
                self.stage = 'down'
                self.rep_start_time = current_time
                self.min_angle_in_rep = elbow_angle # Start tracking the minimum angle

            # While in the 'down' phase, continuously track the lowest point.
            elif self.stage == 'down':
                self.min_angle_in_rep = min(self.min_angle_in_rep, elbow_angle)
                
                # A rep attempt ends when the user has been 'up' for enough frames.
                if self.up_frames >= self.FRAME_BUFFER:
                    self.stage = 'up'
                    
                    # --- 4. Depth Check (at the end of the rep) ---
                    if self.min_angle_in_rep > self.depth_threshold_angle:
                        feedback_this_frame.append("Go deeper on your push-ups!")
                    else:
                        # Only count the rep if the depth was good.
                        self.rep_count += 1
                    
                    # --- 5. Time of Push-up (Tempo Check) ---
                    rep_duration = current_time - self.rep_start_time
                    if rep_duration < self.rep_too_fast_seconds and self.rep_start_time > 0:
                        feedback_this_frame.append("Slow down your reps!")

                    # Reset trackers for the next rep
                    self.rep_start_time = 0
                    self.min_angle_in_rep = 180

            # Add new, unique feedback to the session log
            unique_feedback_in_log = set(self.feedback_log)
            for item in feedback_this_frame:
                if item not in unique_feedback_in_log:
                    self.feedback_log.append(item)
                    unique_feedback_in_log.add(item)
                    
            return feedback_this_frame

        except KeyError as e:
            return []
        except Exception as e:
            return [f"An error occurred: {e}"]

# --- Example Usage ---
if __name__ == '__main__':
    json_file_path = 'goodpushup_training_data.json'
    workout_data = None

    print(f"--- Attempting to load workout data from '{json_file_path}' ---")
    
    try:
        with open(json_file_path, 'r') as f:
            workout_data = json.load(f)
        print("[SUCCESS] Loaded workout data.")
    except FileNotFoundError:
        print(f"\n[ERROR] The file '{json_file_path}' was not found.")
        print("Please ensure the JSON file is in the same directory as this script.")
        exit() # Exit the script if the file doesn't exist
    except json.JSONDecodeError:
        print(f"\n[ERROR] The file '{json_file_path}' is not a valid JSON file.")
        print("Please check the file for formatting errors.")
        exit()

    # Instantiate the specific push-up analyzer
    analyzer = PushupAnalyzer()
    
    print("\n--- Processing a full workout clip frame-by-frame ---")

    # Loop through each frame in the provided data
    if workout_data and "frames" in workout_data:
        for i, frame in enumerate(workout_data["frames"]):
            feedback = analyzer.process_frame(frame)
            if feedback:
                print(f"Frame {i+1}: Feedback -> {feedback}")
    else:
        print("[WARNING] No frames found in the JSON data to process.")


    print("\n--- Workout Complete ---")
    print(f"Total Reps Detected: {analyzer.rep_count}")
    print(f"Final logged feedback for the session: {analyzer.feedback_log}")

File: pushup_analyzer.swift
import Foundation
import simd // For efficient vector math

// --- Step 1: Define Codable Structs to Match Your JSON Data ---

/// Represents a single 3D landmark with visibility.
struct Landmark: Codable {
    let x: Double
    let y: Double
    let z: Double
    let visibility: Double
}

/// Represents a single frame of data from your pose estimation model.
struct FrameData: Codable {
    let relativeTime: Double
    let landmarks: [String: Landmark]

    enum CodingKeys: String, CodingKey {
        case relativeTime = "relative_time"
        case landmarks
    }
}

// --- Step 2: The Main Analyzer Class ---

class PushupAnalyzer {

    // MARK: - Public Properties
    private(set) var repCount = 0
    private(set) var feedbackLog: [String] = []

    // MARK: - Private State Tracking
    private var stage: String = "up"
    private var lastFeedback: String = ""
    private var minAngleInRep: Double = 180.0
    private var frameCount = 0
    
    // Frame Buffer
    private let frameBufferCount = 4
    private var upFrames = 0
    private var downFrames = 0

    // Timers for Tempo
    private var repStartTime: Double = 0.0

    // MARK: - Form Thresholds (Tunable)
    private let depthThresholdAngle: Double = 100.0
    private let repThreshold: Double = 150.0
    private let bodyStraightAngleMin: Double = 150.0
    private let elbowFlareAngleMax: Double = 80.0
    private let repTooFastSeconds: Double = 1.0

    // MARK: - Public Methods

    /// Processes a single frame of landmark data and returns feedback.
    /// - Parameter frameData: The parsed data for the current frame.
    /// - Returns: An array of feedback strings for the current frame.
    func processFrame(frameData: FrameData) -> [String] {
        frameCount += 1
        let landmarks = frameData.landmarks
        let currentTime = frameData.relativeTime
        var feedbackThisFrame: [String] = []

        do {
            // --- Get key landmarks ---
            guard
                let shoulder = landmarks["right_shoulder"],
                let elbow = landmarks["right_elbow"],
                let wrist = landmarks["right_wrist"],
                let hip = landmarks["right_hip"],
                let ankle = landmarks["right_ankle"]
            else {
                throw AnalyzerError.missingLandmark("Required landmark not found.")
            }

            // --- Calculate all relevant angles ---
            let elbowAngle = calculateAngle(a: shoulder, b: elbow, c: wrist)
            let bodyAngle = calculateAngle(a: shoulder, b: hip, c: ankle)
            let elbowFlareAngle = calculateAngle(a: hip, b: shoulder, c: elbow)
            
            // --- 1. Body Straightness Check ---
            if bodyAngle < bodyStraightAngleMin {
                let feedback = "Keep your back straight!"
                if lastFeedback != feedback {
                    feedbackThisFrame.append(feedback)
                    lastFeedback = feedback
                }
            } else {
                lastFeedback = ""
            }

            // --- 2. Elbow Flare Check ---
            if elbowFlareAngle > elbowFlareAngleMax {
                feedbackThisFrame.append("Tuck your elbows in a bit!")
            }

            // --- 3. Rep Counting, Depth, and State Logic with Frame Buffer ---
            if elbowAngle < repThreshold {
                downFrames += 1
                upFrames = 0
            } else {
                upFrames += 1
                downFrames = 0
            }

            // A rep begins when the user has been 'down' for enough frames.
            if stage == "up" && downFrames >= frameBufferCount {
                stage = "down"
                repStartTime = currentTime
                minAngleInRep = elbowAngle // Start tracking
            }
            // While in the 'down' phase...
            else if stage == "down" {
                minAngleInRep = min(minAngleInRep, elbowAngle)

                // A rep attempt ends when the user has been 'up' for enough frames.
                if upFrames >= frameBufferCount {
                    stage = "up"

                    // --- 4. Depth Check (at the end of the rep) ---
                    if minAngleInRep > depthThresholdAngle {
                        feedbackThisFrame.append("Go deeper on your push-ups!")
                    } else {
                        repCount += 1 // Only count good reps
                    }
                    
                    // --- 5. Tempo Check ---
                    let repDuration = currentTime - repStartTime
                    if repDuration < repTooFastSeconds && repStartTime > 0 {
                        feedbackThisFrame.append("Slow down your reps!")
                    }

                    // Reset for the next rep
                    repStartTime = 0
                    minAngleInRep = 180.0
                }
            }
            
            // Add new, unique feedback to the session log
            for item in feedbackThisFrame {
                if !feedbackLog.contains(item) {
                    feedbackLog.append(item)
                }
            }
            
            return feedbackThisFrame

        } catch {
            return ["An error occurred: \(error.localizedDescription)"]
        }
    }
    
    /// Resets the analyzer to its initial state for a new workout session.
    func reset() {
        repCount = 0
        feedbackLog = []
        stage = "up"
        lastFeedback = ""
        minAngleInRep = 180.0
        frameCount = 0
        upFrames = 0
        downFrames = 0
        repStartTime = 0.0
    }

    // MARK: - Private Helpers

    /// Calculates the angle between three landmarks in 2D space (x, y).
    private func calculateAngle(a: Landmark, b: Landmark, c: Landmark) -> Double {
        // Use SIMD for efficient 2D vector operations
        let pA = SIMD2<Double>(a.x, a.y)
        let pB = SIMD2<Double>(b.x, b.y)
        let pC = SIMD2<Double>(c.x, c.y)

        let ba = pA - pB
        let bc = pC - pB
        
        // Calculate cosine of the angle using the dot product
        let cosineAngle = simd_dot(ba, bc) / (simd_length(ba) * simd_length(bc))
        
        // Use acos to get the angle in radians, then convert to degrees
        let angle = acos(cosineAngle)
        return angle * (180.0 / .pi)
    }
    
    enum AnalyzerError: Error, LocalizedError {
        case missingLandmark(String)
        var errorDescription: String? {
            switch self {
            case .missingLandmark(let message):
                return message
            }
        }
    }
}